scheduler/generic_sched_test.go

```diff
@@ -2029,17 +2029,29 @@ func TestServiceSched_JobModify_Canaries(t *testing.T) {
 	if plan.Deployment == nil {
 		t.Fatalf("bad: %#v", plan)
 	}
-	state, ok := plan.Deployment.TaskGroups[job.TaskGroups[0].Name]
-	if !ok {
-		t.Fatalf("bad: %#v", plan)
-	}
+
+	// Ensure local state was not altered in scheduler
+	staleState, ok := plan.Deployment.TaskGroups[job.TaskGroups[0].Name]
+	assert.True(t, ok)
+
+	assert.Equal(t, 0, len(staleState.PlacedCanaries))
+
+	ws := memdb.NewWatchSet()
+
+	// Grab the latest state
+	deploy, err := h.State.DeploymentByID(ws, plan.Deployment.ID)
+	assert.NoError(t, err)
+
+	state, ok := deploy.TaskGroups[job.TaskGroups[0].Name]
+	assert.True(t, ok)
+
 	if state.DesiredTotal != 10 && state.DesiredCanaries != desiredUpdates {
```

Comment:
I'd suggest using `require` here - unlike `assert`, `require` terminate on failure and avoids panics/null deference (e.g. consider when `ok` is false in the test.

Also, you can replace this conditional and the one below with `require.Equal` too.